bytecomp/bytegen.ml

```diff
@@ -345,11 +345,11 @@ let comp_primitive p args =
   | Pfloatcomp Cgt -> Kccall("caml_gt_float", 2)
   | Pfloatcomp Cle -> Kccall("caml_le_float", 2)
   | Pfloatcomp Cge -> Kccall("caml_ge_float", 2)
-  | Pstringlength -> Kccall("caml_ml_string_length", 1)
-  | Pstringrefs -> Kccall("caml_string_get", 2)
-  | Pstringsets -> Kccall("caml_string_set", 3)
-  | Pstringrefu -> Kgetstringchar
-  | Pstringsetu -> Ksetstringchar
+  | Pstringlength | Pbyteslength -> Kccall("caml_ml_string_length", 1)
```

Comment:
The problem is that this partial approach breaks the current coherence.  Today, the runtime system and compiler do not make any difference between strings and bytes.  With your current PR, some C primitives become specialized to bytes/strings, but not all of them.  I can see that this is useful for Bucklescript (which works at the lambda level), but it is not enough for js_of_ocaml and it leaves the system in an inconsistent state.

Formally you don't change the bytecode/native generation, but in practice, the same source code will be compiled differently because of the changes to the non-% primitives in the stdlib.  This means that js_of_ocaml will have to create aliases for its implementation of primitives, without being able to benefit from the bytes/string split.

If you want to split the proposal in two, the natural steps would rather be:
 
1. Do not touch the runtime system at all.  Only introduce some new internal %-primitives (forgetting the strings/bytes distinction when mapping them to C calls) and use them in the stdlib.

2. Make the runtime system aware of the distinction and duplicate all C primitives.  Adapt the remaining non-% primitives in the stdlib and the code generator to target the new primitives.

Your current proposal is between 1 and 2, just enough for Bucklescript, but I think it is worth the extra effort now to do the full 2.